### PR TITLE
feat: Add more thorough setup/teardown hooks

### PR DESCRIPTION
**Pull request by hassila**
## Description
Adding support for global, shared and local setup/teardown hooks in a systemic manner.

Fixes https://github.com/ordo-one/package-benchmark/issues/142

## How Has This Been Tested?

Manual testing, added sample.

## Minimal checklist:

- [x] I have performed a self-review of my own code 
- [x] I have added `DocC` code-level documentation for any public interfaces exported by the package
- [ ] I have added unit and/or integration tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: Sources/Benchmark/Benchmark.swift [hunk 1]
```diff
@@ -41,12 +41,23 @@ public final class Benchmark: Codable, Hashable {
 
     /// Alias for closures used to hook into setup / teardown
     public typealias BenchmarkHook = () async throws -> Void
+    public typealias BenchmarkSetupTeardownHook = BenchmarkHook
+
+    #if swift(>=5.8)
+        @_documentation(visibility: internal)
+    #endif
+    public static var startupHook: BenchmarkSetupTeardownHook? // Should be removed when going to 2.0, just kept for API compatiblity
```
Comment (dimlio) [inline Sources/Benchmark/Benchmark.swift:49]: Add a warning about deprecated API?

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 2]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
 
 There are a number of convenience methods in `BenchmarkThreshold+Defaults.swift`.
 
+### Customize startup/teardown code for benchmarks
+
+There are multiple ways to setup shared benchmark setup/teardown code.
+
+First off, one can easily setup global shared state that can be resued by just defining it in the closure:
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:223]:
```suggestion
First off, one can easily setup global shared state that can be reused by just defining it in the closure:
```

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 3]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+    // Some work to measure here, use mySharedSetup
+    }
+
+    Benchmark("Minimal benchmark 2") { benchmark in
+    // Some work to measure here, use mySharedSetup here too
+    }
+}
+```
+
+Secondly, one can do have setup/teardown closures that are shared across all benchmarks:
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:241]:
```suggestion
Secondly, one can have setup/teardown closures that are shared across all benchmarks:
```
?

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 4]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+
+```swift
+import Benchmark
+
+let benchmarks = {
+    Benchmark.setup = { print("global setup closure, used for all benchmarks") }
+    Benchmark.teardown = { print("global teardown closure, used for all benchmarks") }
+
+    Benchmark("Minimal benchmark") { benchmark in
+    // Some work to measure here, use mySharedSetup
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:251]: No `mySharedSetup` here.

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 5]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+
+func teardownFunction() {
+}
+
+let benchmarks = {
+
+// only shared setup
+Benchmark("Minimal benchmark", 
+    configuration: .init(setup: setupFunction, teardown: teardownFunction)) { benchmark in 
+// Some work to measure here, use mySharedSetup
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:276]:
```suggestion
// Some work to measure here
```

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 6]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+func setupFunction() {
+}
+
+func teardownFunction() {
+}
+
+let benchmarks = {
+
+// only shared setup
+Benchmark("Minimal benchmark",
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:274]: Indent?

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 7]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+
+Finally, one can have setup/teardown closures that are specific to a given benchmark
+
+```swift
+import Benchmark
+
+let benchmarks = {
+
+// only shared setup
+Benchmark("Minimal benchmark",
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:294]: Indent?

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 8]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+Finally, one can have setup/teardown closures that are specific to a given benchmark
+
+```swift
+import Benchmark
+
+let benchmarks = {
+
+// only shared setup
+Benchmark("Minimal benchmark", 
+configuration: .init(setup: setupFunction, teardown: teardownFunction)) { benchmark in
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:295]: Add `setupFunction()` and `teardownFunction()` stubs?

File: Sources/Benchmark/Documentation.docc/WritingBenchmarks.md [hunk 9]
```diff
@@ -216,6 +216,99 @@ Benchmark.defaultConfiguration = .init(...)
...
+```
+
+Finally, one can have setup/teardown closures that are specific to a given benchmark
+
+```swift
+import Benchmark
+
+let benchmarks = {
+
+// only shared setup
```
Comment (dimlio) [inline Sources/Benchmark/Documentation.docc/WritingBenchmarks.md:293]: Doesn't seem to be valid here, or?

**Review by dimlio — APPROVED** [review]

File: Sources/Benchmark/Benchmark.swift (hunk 1 above)
Comment (hassila) [reply Sources/Benchmark/Benchmark.swift:49]: Yeah, I did first do deprecation annotation, but then everyone gets it (even people don't using the hook) which is bad, as we still need to call it, so I just hid it from the documentation so there'll be no new takes on the obsolete naming and then we'll do a break someday...

**Pull request merged**